### DISCUSSION TITLE
How to add audio to stream using .m3u h264/aac rtspsrc

### DISCUSSION THREAD

**Discussion by itsmeaksingh** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#971] [Q&A] [answered]
Hi,

I have used the following command for streaming..

gst-launch-1.0 -v rtspsrc location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=TRUE ! rtph264depay ! h264parse ! kvssink access-key=<Access-key> secret-key=<secret-key> stream-name="ExampleStream" storage-size=512 

And my streaming is successfully displayed on the AWS kinesis-video-streams but i want to listen my audio too. but noting work for me, For audio purpose i have used this command

gst-launch-1.0 -v rtspsrc location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=TRUE ! rtph264depay ! h264parse ! kvssink access-key=<Access-key> secret-key=<secret-key> stream-name="ExampleStream" storage-size=512 alsasrc device=hw:1,0 ! audioconvert ! avenc_aac ! queue ! sink.

can enyone help so i can stream my audio and video. 

![image (1)](https://user-images.githubusercontent.com/52384184/189661140-1cf55d80-69a0-4481-b07b-4322d720e97f.png)
![image](https://user-images.githubusercontent.com/52384184/189661181-866e2766-ae37-4968-8047-b8d06a4a09be.png)

**Comment by hassanctech** [comment]
If you're receiving audio and video from RTSP you will need to split the pipeline and have depayer for h264 as well as for the audio (for example: https://gstreamer.freedesktop.org/documentation/rtp/rtpac3depay.html?gi-language=c), then you need to mux these together to a single stream and then send to kvssink.

Here's and example from someone else who tried to do this, if you follow this it should work for you!  If it does not please re-open: https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/issues/687#issuecomment-807859411

**Comment by itsmeaksingh** [comment]
Hi,

i have check all the solution but nothing worked for me,

Live RTSP link:  rtsp://192.168.68.114:5554/playlist.m3u

# Working without audio
gst-launch-1.0 -v rtspsrc location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=TRUE ! rtph264depay ! h264parse ! kvssink access-key="access-key" secret-key="secret-key" stream-name="ExampleStream" storage-size=512 


----- For audio purose used commands:

# First command for audio  # not working

gst-launch-1.0 rtspsrc name=src location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=true ! rtph264depay ! h264parse ! kvssink name=sink access-key="access-key" secret-key="secret-key" stream-name="ExampleStream" storage-size=1024 aws-region="us-east-2" frame-timecodes=false src. ! rtppcmadepay ! alawdec ! audioresample ! audioconvert ! voaacenc ! aacparse ! queue ! sink.

![first_command](https://user-images.githubusercontent.com/52384184/189879695-32546756-fd78-467a-bbb3-661457288cea.png)
![first_command_part_2](https://user-images.githubusercontent.com/52384184/189879811-a113049a-31a9-4c43-8f0f-5e5ce0ea2445.png)


# Second command for audio  # not working

gst-launch-1.0 rtspsrc name=src location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=true ! rtph264depay ! h264parse ! kvssink name=sink access-key="access-key" secret-key="secret-key" stream-name="ExampleStream" storage-size=1024 aws-region="us-east-2" frame-timecodes=false src. ! rtpmpadepay ! mpg123audiodec ! audioresample ! audioconvert ! voaacenc ! aacparse ! queue ! sink.

![second_command](https://user-images.githubusercontent.com/52384184/189880193-603bfa4e-943e-4737-8025-a7eb6bfce060.png)
![second_command_part_2](https://user-images.githubusercontent.com/52384184/189880907-54742dd1-2e9c-467e-95d0-d393065e130f.png)


# third command for audio  # not working

gst-launch-1.0 rtspsrc name=src location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=true ! rtph264depay ! h264parse ! kvssink name=sink access-key="access-key" secret-key="secret-key" stream-name="ExampleStream" storage-size=1024 aws-region="us-east-2" frame-timecodes=false src. ! rtpaacdepay ! queue ! sink.

![image](https://user-images.githubusercontent.com/52384184/189957321-e21b4d6f-6ee7-473b-af81-4b957567e284.png)



Can you please re-open this issue!

Thanks

**Comment by itsmeaksingh** [comment]
Hi, 

Video:
  Codec: h264
  Resolution: 852 X 480
  Bitrate: 1083 kbps VBR

Audio:
  Codec: aac stereo
  Sample rate: 48000
  Bitrate: 67 kbps VBR

I have using these configuration for my rtsp but my audio is not coming on aws kinesis-video-stream. I don't know much about gst-launch-1.0 library, can you please share with me complete command. 

Thanks

**Comment by javieralcantara** [comment]
Having the same issue here. Did anyone solve it?

**Comment by jdelapla** [comment, marked as answer]
Hello, this is a pure gstreamer issue unrelated to the internals of kvssink. Your pads aren't connecting between your rtspsrc and rtpPCMAdepay elements.

Please add a `queue` element between your src and rtpXXXXdepay(s) when doing a tee like this.

`gst-launch-1.0 rtspsrc name=src location="rtsp://192.168.68.114:5554/playlist.m3u" short-header=true ! queue ! rtph264depay ! h264parse ! kvssink name=sink access-key="access-key" secret-key="secret-key" stream-name="ExampleStream" storage-size=1024 aws-region="us-east-2" frame-timecodes=false src. ! queue ! rtppcmadepay ! alawdec ! audioresample ! audioconvert ! voaacenc ! aacparse ! queue ! sink.`

Something like that should work.
Very similar to this issue someone had with a/v pipelines.
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/issues/1014

**Comment by jdelapla** [comment]
OR your audio codec isn't compatible with rtppcmadepay and you need to select a different depayer.

**Reply by javieralcantara** [reply]
I tried the command @jdelapla suggested without luck. The codec is set to AAC on the camera side.

```
Progress: (open) Opening Stream
Progress: (connect) Connecting to rtsp://xxx:xxxxxxxxxx:554
Progress: (open) Retrieving server options
Progress: (open) Retrieving media info
Progress: (request) SETUP stream 0
Progress: (request) SETUP stream 1
Progress: (open) Opened Stream
Setting pipeline to PLAYING ...
New clock: GstSystemClock
Progress: (request) Sending PLAY request
Progress: (request) Sending PLAY request
Progress: (request) Sent PLAY request
WARNING: from element /GstPipeline:pipeline0/GstRTSPSrc:src: Could not read from resource.
Additional debug info:
gstrtspsrc.c(5280): gst_rtspsrc_loop_udp (): /GstPipeline:pipeline0/GstRTSPSrc:src:
The server closed the connection.
WARNING: from element /GstPipeline:pipeline0/GstRTSPSrc:src: Could not read from resource.
Additional debug info:
gstrtspsrc.c(5427): gst_rtspsrc_reconnect (): /GstPipeline:pipeline0/GstRTSPSrc:src:
Could not receive any UDP packets for 5.0000 seconds, maybe your firewall is blocking it. Retrying using a tcp connection.
WARNING: from element /GstPipeline:pipeline0/GstRTSPSrc:src: Delayed linking failed.
Additional debug info:
./grammar.y(510): gst_parse_no_more_pads (): /GstPipeline:pipeline0/GstRTSPSrc:src:
failed delayed linking some pad of GstRTSPSrc named src to some pad of GstQueue named queue1
[INFO ] [23-02-2023 18:09:24:409.714 GMT] kinesisVideoStreamFormatChanged(): Stream format changed.
```

**Reply by jdelapla** [reply]
You would have to change the rtspsrc location to an rtspsrc location you know is providing a stream. Your rtspsrc error is stating it's not receiving any packets.

**Reply by javieralcantara** [reply]
No, that RTSP stream definitely exists. I can get video out of it using FFMPEG but not using gst-launch.

**Comment by hassanctech** [comment]
If you're using `aac` audio the depayer you're looking for is this one:

```
rtp:  rtpmp4adepay: RTP MPEG4 audio depayloader
```

**Reply by javieralcantara** [reply]
Please could you provide an example? I am trying this but I keep getting an error.

## Working example without audio:
```
AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY AWS_DEFAULT_REGION=$AWS_REGION gst-launch-1.0 rtspsrc location=$STREAM_URL ! rtph264depay ! h264parse ! kvssink stream-name="LATEST_TEST" storage-size=512**
```

## Not working with audio:
```
AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY AWS_DEFAULT_REGION=$AWS_REGION gst-launch-1.0 rtspsrc name=src location=$STREAM_URL short-header=true ! queue ! rtph264depay ! h264parse ! kvssink name=sink stream-name="ExampleStream" storage-size=1024 frame-timecodes=false src. ! queue ! rtppcmadepay ! alawdec ! audioresample ! audioconvert ! voaacenc ! aacparse ! queue ! sink.
```
Error:
```
WARNING: from element /GstPipeline:pipeline0/GstRTSPSrc:src: Delayed linking failed.
Additional debug info:
./grammar.y(506): gst_parse_no_more_pads (): /GstPipeline:pipeline0/GstRTSPSrc:src:
failed delayed linking some pad of GstRTSPSrc named src to some pad of GstQueue named queue1
[INFO ] [27-02-2023 20:10:32:069.168 GMT] kinesisVideoStreamFormatChanged(): Stream format changed.
```

**Reply by javieralcantara** [reply]
Okay the command bellow seems to work but it really does not. I get error code `0x52000085` after a few seconds.

```
gst-launch-1.0 rtspsrc name=src location=$STREAM_URL short-header=true ! queue ! rtph264depay ! h264parse ! kvssink name=sink stream-name="ExampleStream" storage-size=1024 frame-timecodes=false src. ! queue ! rtpmp4gdepay ! aacparse ! queue ! sink.
```

According to [this](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-java/issues/113) there is an encoder issue. This is where I get lost. Anyone with experience know what is going on?
